components/camera/videosource/logging/logger.go

```diff
@@ -0,0 +1,248 @@
+				return
+			case info := <-l.infoCh:
+				l.write(info.title, info.m)
+			case <-ticker.C:
+				l.captureV4L2info()
+			}
+		}
+	})
+}
+
+// Done signals to the logger there's one less entity dependent on this logger.
+//
+// It will block until the logger has shutdown all background threads if and only if it is the last entity watched.
+// This method is thread safe. See Add.
+func (l *Logger) Done() {
+	if j := l.nActive.Add(-1); j == 0 {
+		l.cancel()
+		<-l.doneCh
+	}
+}
+
+// Log adds to infoMap to the log file under the given title. It only adds entries from the infoMap not already present
+// under the given title. This method is thread safe.
+func (l *Logger) Log(title string, m InfoMap) {
+	if !l.isRunning.Load() {
+		panic("please add a dependency before attempting to log.")
+	}
+
+	l.infoCh <- info{title, m}
+}
+
+func (l *Logger) captureV4L2info() {
+	v4l2Info := make(InfoMap)
+	v4l2Compliance := make(InfoMap)
+
+	// Video capture and overlay devices' minor numbers range from [0,63]
+	// https://www.kernel.org/doc/html/v4.16/media/uapi/v4l/diff-v4l.html
+	for n := 0; n < 63; n++ {
+		path := fmt.Sprintf("/dev/video%d", n)
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			continue
+		}
+
+		v4l2Info[path] = l.runCommand("v4l2-ctl", "--device", path, "--all")
+		v4l2Compliance[path] = l.runCommand("v4l2-compliance", "--device", path)
+	}
+
+	l.write("v4l2 info", v4l2Info)
+	l.write("v4l2 compliance", v4l2Compliance)
+}
+
+func (l *Logger) init() {
+	header := createHeader("system information", 0)
+	l.Logger.Infoln(header)
+
+	out := l.runCommand("uname", "--all")
+	l.Logger.Infoln(out)
+}
+
+func (l *Logger) runCommand(name string, args ...string) string {
+	cmd := exec.Command(name, args...)
+	//nolint:errcheck
+	out, _ := cmd.CombinedOutput()
+	return string(out)
+}
+
+func createHeader(title string, _ int) string {
+	// TODO: add different headers for different levels (_ int)
+	lineLen := 80
+	padding := lineLen - len(title) - 4
+
+	return fmt.Sprintf("#%s %s %s#\n",
+		strings.Repeat("=", padding/2),
+		title,
+		strings.Repeat("=", int(math.Ceil(float64(padding)/2))),
+	)
+}
+
+func mapDiff(oldM, newM InfoMap) (added, removed, modified InfoMap) {
+	added, removed, modified = make(InfoMap), make(InfoMap), make(InfoMap)
+
+	for k, v := range newM {
+		if _, ok := oldM[k]; !ok {
+			added[k] = v
+		}
+	}
+
+	for k, v := range oldM {
+		if _, ok := newM[k]; !ok {
+			removed[k] = v
+		}
+	}
+
+	for k, oldV := range oldM {
+		if newV, ok := newM[k]; ok && strings.Compare(newV, oldV) != 0 {
+			modified[k] = fmt.Sprintf("%s -> %s", oldV, newV)
+		}
+	}
+
+	return
+}
+
+func (l *Logger) writeMap(title string, m InfoMap) {
+	header := createHeader(title, 1)
+	l.Logger.Infoln(header)
+
+	for k, v := range m {
+		l.Logger.Infof("%s: %s\n", k, v)
+	}
+}
+
+func (l *Logger) write(title string, m InfoMap) {
+	defer func() {
+		l.prevInfo[title] = m
+	}()
+
+	added, removed, modified := mapDiff(l.prevInfo[title], m)
```

Comment:
I think when the file rolls over we need to write out the full map. Or we could simplify it by just writing out the full map with some frequency (or not doing map diffing at all).

Reply:
Ah, good point. The map diffing was solely to save space and help with readability. I'll make sure to fix this.